cities/queries/causal_insight.py

```diff
@@ -0,0 +1,413 @@
+        # wrt to actual observed outcomes rather than predicting outcomes themselves
+        # effectively keeping the noise fixed and focusing on a counterfactual claim
+
+        # TODO possible delete in the current strategy deemed uncontroversial
+        # else:
+        #     if not isinstance(intervened_value, torch.Tensor):
+        #         intervened_value = torch.tensor(intervened_value, device=self.device)
+        #     intervened_expanded = intervened_value.expand_as(self.data['t'])
+
+        #     with MultiWorldCounterfactual(first_available_dim=-6) as mwc:
+        #         with do(actions = dict(T = intervened_expanded)):
+        #                 self.predictive = pyro.infer.Predictive(model=self.model, guide=self.guide,
+        #                         num_samples=self.num_samples, parallel=True)
+        #                 self.samples = self.predictive(*self.model_args)
+        #     self.mwc = mwc
+
+    def generate_tensed_samples(self):
+        self.tensed_samples = {}
+        self.tensed_tau_samples = {}
+
+        for shift in [1, 2, 3]:
+            self.load_guide(shift)
+            self.generate_samples()
+            self.tensed_samples[shift] = self.samples
+            self.tensed_tau_samples[shift] = (
+                self.samples["weight_TY"].squeeze().detach().numpy()
+            )
+
+    def get_fips_predictions(self, fips, intervened_value, year=None):
+        self.fips = fips
+        self.intervened_value = intervened_value
+        # start with the latest year possible by default
+        if year is None:
+            year = self.data["years_available"][-1]
+        assert year in self.data["years_available"]
+
+        self.year = year
+
+        # find years for prediction
+        outcome_years = self.data["outcome_years"]
+        year_id = [int(x) for x in outcome_years].index(year)
+        self.year_id = year_id
+
+        self.prediction_years = outcome_years[(year_id) : (year_id + 4)]
+
+        # find fips unit index
+        dg = DataGrabber()
+        dg.get_features_std_wide([self.intervention_dataset, self.outcome_dataset])
+        dg.get_features_wide([self.intervention_dataset])
+        interventions_this_year_original = dg.wide[self.intervention_dataset][str(year)]
+
+        self.intervened_value_original = revert_standardize_and_scale_scaler(
+            self.intervened_value, self.year, self.intervention_dataset
+        )
+
+        # interventions_this_year = dg.std_wide[self.intervention_dataset][str(year)]
+
+        # nearest_intervention_index = min(
+        #     range(len(interventions_this_year)),
+        #     key=lambda i: abs(interventions_this_year[i] - self.intervened_value),
+        # )
+        # self.intervened_value_original = interventions_this_year_original[
+        #     nearest_intervention_index
+        # ]
+
+        self.intervened_value_percentile = round(
```

Comment:
The data processing stuff above can also probably be broken out into its own method.

Reply:
okay, I think part of why the class is kinda annoying is that I could move some of the code to external modules and call them after importing here. Will make smalls steps in this direction, for now made an issue about it.